internal/controller/controlplane/k0s_controlplane_controller.go

```diff
@@ -357,19 +358,27 @@ func (c *K0sController) reconcileMachines(ctx context.Context, cluster *clusterv
 			clusterIsUpdating = true
 			clusterIsMutating = true
 			if kcp.Spec.UpdateStrategy == cpv1beta1.UpdateInPlace {
-				desiredMachineNames[m.Name] = true
+				desiredMachineNamesSlice = append(desiredMachineNamesSlice, m.Name)
 			} else {
 				machineNamesToDelete[m.Name] = true
 			}
 		} else if !matchesTemplateClonedFrom(infraMachines, kcp, m) {
 			clusterIsMutating = true
 			machineNamesToDelete[m.Name] = true
-		} else if activeMachines.Len() > int(kcp.Spec.Replicas)+len(machineNamesToDelete) {
-			machineNamesToDelete[m.Name] = true
 		} else {
-			desiredMachineNames[m.Name] = true
+			desiredMachineNamesSlice = append(desiredMachineNamesSlice, m.Name)
 		}
 	}
+	desiredMachineNames := make(map[string]bool)
+	for i := range desiredMachineNamesSlice {
+		desiredMachineNames[desiredMachineNamesSlice[i]] = true
+	}
+
+	// if it is necessary to reduce the number of replicas even counting the replicas to be eliminated
+	// because they are outdated, we choose the oldest among the valid ones.
+	if activeMachines.Len() > int(kcp.Spec.Replicas)+len(machineNamesToDelete) && len(desiredMachineNamesSlice) > 0 {
```

Comment:
Could you elaborate a bit on why the logic was changed? The result looks more or less the same

Reply:
Great catch!